src/childinfo.c

```diff
@@ -54,34 +60,93 @@ void closeChildInfoPipe(void) {
+        return;
+    }
+
+    if (process_type == CHILD_TYPE_RDB) {
+        server.stat_rdb_cow_bytes = cow_size;
+    } else if (process_type == CHILD_TYPE_AOF) {
+        server.stat_aof_cow_bytes = cow_size;
+    } else if (process_type == CHILD_TYPE_MODULE) {
+        server.stat_module_cow_bytes = cow_size;
+    }
+}
+
+/* Read COW info data from the pipe.
+ * if complete data read into the buffer, process type, copy-on-write type and copy-on-write size
+ * are stored into *process_type, *cow_type and *cow_size respectively and returns 1.
+ * otherwise, the partial data is left in the buffer, waiting for the next read, and returns 0. */
+int readChildInfo(int *process_type, int *cow_type, size_t *cow_size) {
+    /* we are using here a static buffer in combination with the server.child_info_nread to handle short reads */
+    static child_info_data buffer;
+    ssize_t wlen = sizeof(buffer);
+
+    /* do not overlap */
+    if (server.child_info_nread == wlen) server.child_info_nread = 0;
+
+    int nread = read(server.child_info_pipe[0],&buffer + server.child_info_nread, wlen - server.child_info_nread);
+    if (nread > 0) {
+        server.child_info_nread += nread;
+    }
+
+    /* we have complete child info */
+    if (server.child_info_nread == wlen) {
+        *process_type = buffer.process_type;
+        *cow_type = buffer.cow_type;
+        *cow_size = buffer.cow_size;
+        return 1;
+    } else {
+        return 0;
+    }
+}
+
 /* Receive COW data from parent. */
 void receiveChildInfo(void) {
     if (server.child_info_pipe[0] == -1) return;
-    ssize_t wlen = sizeof(server.child_info_data);
-    if (read(server.child_info_pipe[0],&server.child_info_data,wlen) == wlen &&
-        server.child_info_data.magic == CHILD_INFO_MAGIC)
-    {
-        if (server.child_info_data.process_type == CHILD_TYPE_RDB) {
-            server.stat_rdb_cow_bytes = server.child_info_data.cow_size;
-        } else if (server.child_info_data.process_type == CHILD_TYPE_AOF) {
-            server.stat_aof_cow_bytes = server.child_info_data.cow_size;
-        } else if (server.child_info_data.process_type == CHILD_TYPE_MODULE) {
-            server.stat_module_cow_bytes = server.child_info_data.cow_size;
-        }
+
+    int process_type;
+    int cow_type;
+    size_t cow_size;
+    if (readChildInfo(&process_type, &cow_type, &cow_size)) {
+        updateChildInfo(process_type, cow_type, cow_size);
+    }
+}
+
+/* Receive last COW data from parent. */
+void receiveLastChildInfo(void) {
+    if (server.child_info_pipe[0] == -1) return;
+
+    /* drain the pipe, and check last read data (if exist) */
+    int child_info_available = 0;
+    int process_type;
+    int cow_type;
+    size_t cow_size;
+
+    while (readChildInfo(&process_type, &cow_type, &cow_size) > 0) {
+        child_info_available = 1;
+    }
+
+    if (child_info_available) {
+        updateChildInfo(process_type, cow_type, cow_size);
```

Comment:
it is common for functions with output parameters to reset them on entrance, so that when they return an error, they don't contain random values.
if readChildInfo was doing that, then the previous code would break, since it relies on the values that returned from the pre-ultimate call.
i think this is a better patter (considering that updateChildInfo is cheap)
```suggestion
    while (readChildInfo(&process_type, &cow_type, &cow_size) > 0) {
        updateChildInfo(process_type, cow_type, cow_size);
```